### ISSUE TITLE
discussion: persistent storage strategies [cookies, indexeddb + local dom]

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#188]
Currently we have `network.cookie.cookieBehavior` set as 2 with a mention that `You can set exceptions under site permissions or use an extension`

0=allow all 1=allow same host 2=disallow all 3=allow 3rd party if it already set a cookie

We could change this to 1 as it would fix some underlying issues (but maybe raise some).
- Blocking cookies also blocks local DOM storage, this causes Extensions that use local storage to break (1 example so far that demonstrates this, eCleaner, although he could fix this)
- uMatrix (which not all users would want to or can use) can be used to default block all cookies (which negates the need to use site exceptions
- FPI isolates cookies, in fact Arthur E from TOR has said he wants to enable ALL cookies including 3rd party in TBB in the future.
- The downside to FPI and cookies is that any cookie extension cannot touch FPI cookies - hence default block all in uMatrix. You can only manually delete cookies (but I am sure they can be cleared on shutdown) - see [1381197](https://bugzilla.mozilla.org/show_bug.cgi?id=1381197)
- If cookies are allowed, you will also get a SB google cookie. It will show in your cookies UI, and you can manually delete it, but this a priviledged cookie in a separate jar and no extension is ever going to be allowed to delete it (Francois argued that it should be accessible to extensions, but they said NO - the bastards). To date I do not nor ever have had a google cookie. I do not know if uMatrix can block this instance (would be nice to know). May need a global google no cookie rule. See [1362834](https://bugzilla.mozilla.org/show_bug.cgi?id=1362834)
- Just to add .. PB mode cookies are protected from extensions as well
- If cookies are allowed, this allows all sites to use local DOM storage. I am almost certain that uMatrix's blocking of the cookie will have no affect on access to local storage - so that then means it needs cleaning
- Cookie extensions cannot clean local dom storage (yet)
- Test page: http://dev-test.nemikor.com/web-storage/support-test/  (thanks @2glops )

Why can't ALL cookies be accessible in ALL modes (PB, FPI, TP+SB), as well as local dom storage

So, as you can see .. downsides, but upsides

Class, discuss!

**Comment by Thorin-Oakenpants** [comment]
With FPI and PB mode - cookie addons are useless for cleaning. They are OK for blocking (see uMatrix) - so enabling cookies is FINE - my concern is the local storage

**Comment by Atavic** [comment]
> 3=allow 3rd party if it already set a cookie

Does it still work? Seems abandonware to me.

**Comment by Thorin-Oakenpants** [comment]
^^ no idea, probably means nothing in the context of FPI as per domain (origin) will cause multiples of it. We're not even considering it anyway, so who cares :)

**Comment by Thorin-Oakenpants** [comment]
FYI:
https://bugzilla.mozilla.org/show_bug.cgi?id=1362834 - protected SB google cookie
https://bugzilla.mozilla.org/show_bug.cgi?id=565965 - FPI & cookies

**Comment by Thorin-Oakenpants** [comment]
PS: I suggest if you are using FPI to delete all your cookies and let them get re-created so they have the correct attributes

**Comment by Just-me-ghacks** [comment]
This is what I use:
user_pref("network.cookie.cookieBehavior", 1);
user_pref("dom.storage.enabled", false);

I don't block ALL cookies with uMatrix. Should I?

Occasionally I have to enable dom storage but after that I delete my entire Firefox folder.

**Comment by Thorin-Oakenpants** [comment]
> I haven't seen any "SB google cookie". Ever. I have all the SB stuff disabled. Where should I check for this cookie?

AFAIK, if you have it, it should be visible under about:preferences#privacy > View Cookies - see https://bugzilla.mozilla.org/show_bug.cgi?id=1362834

> I don't block ALL cookies with uMatrix. Should I?

Why not? Use uM exceptions. Currently I have to allow an exception in both FF's site permissions AND uM, so setting cookies as 1st party removes a step - that is NOT the issue. The question is not really about cookies but about dom storage (which is in effect controlled by FF's internal cookie setting).

When I ditched Cookie Controller, I removed all my cookies, removed all cookie exceptions, FPI is on, and recreated site exceptions, created exceptions in uM, and recreated the cookies by visiting the sites (this to make sure all cookies are new with the right attributes etc) - seriously .. I now have a grand total of 7 cookies and 1 session cookie if it exists) = 8 sites total. Why bother allowing cookies by default. A whitelist saves so much BS. A block all also means much better control over them due to FPI & PB mode issues re extensions.

---

The equation is (remembering that the cookie setting also controls dom storage)

Either
1.  cookie :x: dom storage :white_check_mark: [CURRENT]
   * means simply allow site exceptions (eg allow, allow for session) for sites that need dom (I have to admit I see very very little of this requirement with my web browsing) or a cookie for logins
   * You could still block the cookie itself (eg uMatrix) but the firefox internal pref means dom is granted (pretty sure uM blocking the cookie does not affect local dom = needs testing to confirm)
   * You still need to clear dom (for the few sites it gets set)
2. cookie :white_check_mark: dom storage :white_check_mark: [POSSIBLE]
   * by allowing cookies this allows dom storage, even if cookies are blocked in uM (needs test). This makes websites work = less breakage, but also allows WebExtensions that use local storage to also work (uBo may be one of these) :+1: 
   * you still need to clear dom (but now its all sites, not a few) :-1: 
   * note that FPI means dom storage is isolated which is good, but it's also re-used and can track you when reloading the site. This has always been the case
3.  cookie :x: dom storage :x: 
   * idk
4.  cookie :white_check_mark:  dom storage :x: 
   * idk

Under site prefs is an option to Maintain Online Storage. IF this directly relates to dom storage, then the solution is number 4 - use site exceptions for cookies in uM and FF site exceptions for dom storage to offer the perfect granular control

**Comment by earthlng** [comment]
I'm confused about your use of "dom storage" - What is that? I think you're mixing them up.

What kind of offline storage exists and where is it stored should be the first thing we clarify.

Afaik, there's
* indexedDB (=disabled)
  * <profiledir>/storage/permanent/... (??) I've also seen /storage/temporary/...
* "offline storage" which is apparently != indexedDB ?
  * storage.sqlite ?? storage-sync.sqlite ?? webappstore.sqlite ??
  * about:cache > appcache? disabled by `browser.cache.offline.enable;false` ?
* WE's local.storage which is used to store a WE's settings
  * stored in <profiledir>/browser-extension-data/<WE-id>/storage.js

Is there anything else?

**Comment by Thorin-Oakenpants** [comment]
^^^^^ THIS one

```js
/* 2705: disable DOM (Document Object Model) Storage
 * [WARNING] This will break a LOT of sites' functionality.
 * You are better off using an extension for more granular control ***/
   // user_pref("dom.storage.enabled", false);  <<< ELI5 WHERE THIS LIVES
```

THIS is called local dom storage by almost everything I know

**Comment by Thorin-Oakenpants** [comment]
@fmarier : can you confirm that all identifying information in any SB google cookie i stripped out. I do not have one to look at. Is it a requirement (ToS) that a SB cookie even be used, because I update just fine and I block it

**Comment by 2glops** [comment]
Found that test page from MDN :
http://dev-test.nemikor.com/web-storage/support-test/

**Comment by Thorin-Oakenpants** [comment]
^^ @2glops - did you allow a site exception for cookies for that test site? I disabled uM and uBo and NS temporarily for that site, added a cookie exception (site prefs) and clicked run the tests and nothing seems to happen.

**Comment by 2glops** [comment]
With site exception cookie (about:prefs), that returns :
localStorage: limited to 5101 k characters
sessionStorage: limited to 5101 k characters
globalStorage: not supported

**Comment by 2glops** [comment]
With site exception cookie (about:prefs), that returns :
localStorage: limited to 5101 k characters
sessionStorage: limited to 5101 k characters
globalStorage: not supported

The tests seems to have completed (5 min run) but no dialog to confirm)

**Comment by Thorin-Oakenpants** [comment]
Right, which is exactly what I have been saying, that the FF pref for cookies also controls the local storage doesn't matter if it is global or a site exception.

I got the same result with eCleaner - even though it doesn't need or set a cookie, it requires cookies enabled (0 or 1), as then allows local dom storage, which in turn meant the Web Ext could access it's saved settings

We need a test where a cookie is set and it uses local storage. So we can allow the site a cookie in FF, but block the cookie in uM, and see if local dom is still accessible

**Comment by Thorin-Oakenpants** [comment]
This is interesting: https://addons.mozilla.org/en-US/firefox/addon/shim-storage/ (legacy addon)

> If your disabling localStorage or sessionStorage many sites stops working. Not cause they actually need the database, but cause they expect it to be there.
> 
> This extension allows you to define a list of domains where a shim should be defined so the page will actually work. No data will be stored anywhere outside the page.
> 
> NOTE: This extension has no effect if you don't set dom.storage.enabled to false.

So in effect:
1. Allow cookies in FF (which in effect NOW stops blocking dom storage) but block all cookies in uMatrix by default. uM giving you granular control
2. Deny dom.storage.enabled and use the shim storage extension for granular control

Wonder how hard it would be or if even possible in a WebExt? However, this still doesn't exceptions for allowing Web Extensions local storage AFAIK, which is a problem.

Just about ready to give up. I hope the new about:permissions gives a much more precise and thorough configuration (eg indexedb, dom storage, offline storage, cookies = and none are dependent on others, and extensions are exempt)

**Comment by Atavic** [comment]
Re: https://github.com/ghacksuserjs/ghacks-user.js/issues/188#issuecomment-320470847

3=allow 3rd party had it's own policy (P3P) that was taken down by google, inc. et al.

**Comment by earthlng** [comment]
Hey Pants, don't you get a bunch of error message in the console after startup in 55?
FF also uses indexedDB internally and disabling it might have unforeseen consequences. I've now enabled it again and will use PB windows to block sites from using it. If a site needs JS I'll make sure to open it in PB.

As for "Web Storage" (aka DOM Storage) uMatrix has an option to delete it for blocked hostnames.
I'm using 1 as the cookie value (no 3rd party cookies), indexedDB enabled, dom.storage.enabled=true and clear cookies on shutdown which also deletes sites "localstorage".
testpage: https://mdn.github.io/dom-examples/web-storage/
Information is stored in the **webappsstore.sqlite** file for websites and in the **chromeappsstore.sqlite** for [about:*] ([source](https://support.mozilla.org/en-US/kb/profiles-where-firefox-stores-user-data))

**Comment by Thorin-Oakenpants** [comment]
I dont give a sh*t about console errors :) Why would mozilla provide an indexedDB pref that breaks INTERNAL code.

How can I see the console error messages when I start? EIL5 please.

If we need to allow indexedDB this solves my issue re Stylus replacing Stylish - but there is no way to control indexedDB (and I've never really had any issues with sites being denied)

One way to deal with this is to allow indexedDB in prefs, but deny system permission to the relevant folders - `...\profile\storage\temporary`

**Comment by Thorin-Oakenpants** [comment]
Looking in my `"D:\Portable\FirefoxPortable\Data\profile\storage\permanent\"` (which I assume is FF internals), I see date modified is all like 2 years old - heh

**Comment by earthlng** [comment]
You don't clean up your profile folder very much then, huh? xD

> Why would mozilla provide an indexedDB pref that breaks INTERNAL code.

IDK but they do.

> How can I see the console error messages when I start? EIL5 please.

I have to ELI5 a lot recently ;) start FF then Ctrl+Shift+J

**Comment by Thorin-Oakenpants** [comment]
> You don't clean up your profile folder very much then, huh

I do. There's nothing in there. 
`...\profile\storage\default` and `...\profile\storage\temporary` are empty
`...\profile\storage\permanent` has three sub-dirs: `chrome`, `indexeddb+++fx-devtools` and `moz-safe-about+home` and these are internal, so I don't care about cleaning them

I guess its been about 2 years since I toggled indexedDB on. My profile is well over 4 years old. I keep it clean as I can (removed dead prefs including extension ones).

Maybe with 57 I will start a brand new clean profile.

**Comment by earthlng** [comment]
noticed all the error messages with indexedDB disabled?

**Comment by Thorin-Oakenpants** [comment]
No, because I haven't had time yet, and I am about to go AWOL for 24 hrs. Need a break from you all. I've made a note to check out the errors :)

We need a new strategy to overall deal with cookies (easy) and all the storage types and how to clean them as well as Web Ext storage (and now FF internal storage) and everything is a mess due to restrictions .. almost had enough for today :space_invader: vs :robot:

**Comment by Thorin-Oakenpants** [comment]
I like your stragegy (remember that FPI is on for storage types so storage is really only acting as first party) where clearing cookies on shutdown clears local dom storage. I can live with that - slight PITA - means I always have to log back in to my sites - might keep FF open for longer periods. no big deal

**Comment by Thorin-Oakenpants** [comment]
FYI - nilla 55 profile with indexedDB disabled

![noindexeddb](https://user-images.githubusercontent.com/16656956/29164294-101da666-7e13-11e7-8636-2a64588207e5.png)

AFAIK, this isn't affecting ANYTHING I care about

**Comment by Thorin-Oakenpants** [comment]
FYI:

```
unterminated regular expression literal  Clear_window.opener.user.js:1
```

**Comment by earthlng** [comment]
Are you sure there isn't something wrong with your copy?

> Clear_window.opener.user.js:1

is it really line 1 that's causing the problem??

**Comment by Thorin-Oakenpants** [comment]
damn .. it was that missing `/`  - Wot the F , I swear I edited mine and saved it when told - see #155

```
/ ==UserScript==
```

**Comment by earthlng** [comment]
see.. taking a look at the console from time to time is pretty neat

**Comment by Thorin-Oakenpants** [comment]
From Atavic
> >    3=allow 3rd party if it already set a cookie
> Does it still work? Seems abandonware to me

```js
 0=allow all 1=allow same host 2=disallow all 3=allow 3rd party if it already set a cookie
 [SETTING] Options>Privacy>History>Custom Settings>Accept cookies from sites
```

@Atavic :
- I have Accept cookies from sites - two options, checked/unchecked
- I have Accept 3rd Party Cookies - "Always", "If Visited", and "Never" (which only applies if the above is checked)
- This allows 4 outcomes

- If you uncheck "Accept Cookies" - you will get a `network.cookie.cookieBehavior` = 2
- If you  click "Accept Cookies" (checkbox) and
  * select  "Accept third-party cookies>Always" - you will get a `network.cookie.cookieBehavior` = 0
  * select "Accept third-party cookies>From Visited" - you will get a `network.cookie.cookieBehavior` = 3
  * select "Accept third-party cookies>Never" - you will get a `network.cookie.cookieBehavior` = 1

So that maps nice and correct to what we said in the user.js

**Comment by Thorin-Oakenpants** [comment]
For extensions that "require" cookies (they don't actually need them) such as Stylus, you can set an extension exception by going to about:debugging, selecting the manifest URL and from there adding your exception from the urlbar info icon etc.

It's looking more and more like cookies will need to be unblocked at the core FF level (1st party only), and blocked by eg uMatrix. I am OK with this.... BUT uMatrix will then only block the cookie itself and now every single site will have access to local dom storage (right?). This creates a problem - eg auto-cleaning local dom storage eg via Cookie-AutoDelete - i.e, no cookie, no detection of dom to clear (note that the bugs that prevented cleaning dom have and are being resolved fixed).

@earthlng - can you please tell me, yes or no, the answer to this:
Q: If I allow cookies in FF Options, but block them using uM, can a website still set local dom storage

**Comment by Atavic** [comment]
Gorhill's Efficient blacklisting:

>Blacklisted cookies are not prevented by uMatrix from entering your browser. However they are prevented from leaving your browser , which is what really matters.

(from wiki)

**Comment by Thorin-Oakenpants** [comment]
Ahh, well that changes things - clearly local dom storage is opened up as well. So it then becomes an issue when not in default-deny to take care all all the fallout of cookies + dom and cleaning them out  What a mess. I have loved the last 5 years without having to clean hardly anything.

Maybe the extension exclusion will be enough. I guess I'll find out as I go along. Meanwhile, indexeddb I might put system permission locks on `...\storage\default` and `...\storage\temporary` - although 4 or 5 days into allowing indexeddb, there is still nothing in there

**Comment by earthlng** [comment]
> @earthlng - can you please tell me, yes or no, the answer to this:
> Q: If I allow cookies in FF Options, but block them using uM, can a website still set local dom storage

Yes but you can enable "Delete local storage content set by blocked hostnames".

Mozilla has setup a nice [demo page](https://mdn.github.io/dom-examples/web-storage/) where you can test it yourself.
If you don't allow the cookie for that page in uMatrix you can see in uM's logger that localstorage is blocked: `--	cookie	https://mdn.github.io/{localStorage}`

It's a bit weird though because you can change the values on that page and a normal refresh of the page will still use those values. However when you shift+refresh the page it resets to the red background and the default values. For me that's good enough. But that's with the old non-WE uMatrix! They are still working on allowing WE's to clear localstorage for certain hosts only.

**Comment by Thorin-Oakenpants** [comment]
Can't be bothered reading up and searching, not sure if anyone said indexeddb was cleared on close with anything .. (clue, it's not) - here is a demo page: https://static.raymondcamden.com/demos/2014/feb/7/index.html#/home (not that I tried it)

Here is a [1340511](https://bugzilla.mozilla.org/show_bug.cgi?id=1340511) for `extend browsingData api for single site` - that is the ability to remove site data per domain (cookies, localStorage,  indexedDB,  serviceWorkers, cacheStorage, applicationCache)

Here is [1333050](https://bugzilla.mozilla.org/show_bug.cgi?id=1333050) which is one step of the above, `clearing of IndexedDB in browsingData API`

**Comment by crssi** [comment]
^^The demo page is using default storage in profile/storage/default folder location.
This is persistent and it does not get cleared with any to me known preference or addon and I have all kind of clearing mechanisms in place.
The only avoiding way as I see right now is to delete default and temporary folders and make files with the same names (file name extensionless) to block it.
But in that case adding a node on this demo page doesn't even work as intended.

Is there anyway ATM to clean those folders with a preferences or addon?

**Comment by Thorin-Oakenpants** [comment]
No. That's why I linked to 1333050 ... indexedDB has never really had any access to my knowledge and is a real problem

**Comment by crssi** [comment]
I guessed so. Making those folders read only can be tricky, since FF update could "repair" ACLs on those and better solution would be to remove those and create files with the same name.
I have done so ages back and I can't remember if any site broke because of it... with an exception of here mentioned demo site.

**Comment by earthlng** [comment]
Or you know, you could just use a Private Window for sites you allow to use Javascript

**Comment by crssi** [comment]
The demo site doesn't work in PB. Also I loose control over cookies in PB.

**Comment by Forsaked** [comment]
@crssi Any Outlook Web App would break without IndexDB enabled.

**Comment by crssi** [comment]
@Forsaked see [https://github.com/ghacksuserjs/ghacks-user.js/issues/180#issuecomment-317660941](https://github.com/ghacksuserjs/ghacks-user.js/issues/180#issuecomment-317660941)
;)
And I have IndexDB enabled long time ago, never actually disabled it (only for testing).

**Comment by earthlng** [comment]
Why not? the site works fine for me in PB

**Comment by crssi** [comment]
@earthlng 
no it does not... you made me retest with a clean profile. :)
it opens, but doesn't work as intended... try click on "Add Node" green button on the right... nothing happens in PB and the same result in normal browsing with IndexDB enabled and blocked storage (default) on file level.

**Comment by earthlng** [comment]
Oh, I thought we're talking about this test site: https://mdn.github.io/dom-examples/web-storage/

sorry :)

**Comment by crssi** [comment]
:)
In my configuration with normal browsing FF doesn't remember anything with test site you mentioned without restarting FF.

**Comment by earthlng** [comment]
[Bug 781982](https://bugzilla.mozilla.org/show_bug.cgi?id=781982) - IndexedDB does not function in private browsing mode

**Comment by Thorin-Oakenpants** [comment]
Now that's funny. One of PB mode's purposes is to not leave persistent data on close, and because there is no way for FF to clean indexedDB, they just left disabled it. @fmarier - that's not a solution. Breaking sites in PB mode like that. Don't you get telemetry analysis bringing these things up? I would have expected thousands of complaints that PB mode breaks sites. Is there any telemetry on % of sites that use indexedDB?

**Comment by fmarier** [comment]
> Don't you get telemetry analysis bringing these things up?

We have a policy of not collecting any telemetry in Private Browsing mode. That might change once that RAPPOR system is in place, but right now we know very little about what goes on in Private Browsing.

**Comment by Thorin-Oakenpants** [comment]
Doh! Of course. I like that policy :+1: Don't change it EXCEPT if you do, limit it to Nightly

**Comment by Thorin-Oakenpants** [comment]
Looks like [1388428 - Extend browsingData to support removing localStorage by host](https://bugzilla.mozilla.org/show_bug.cgi?id=1388428) isn't far away, which will allow C-AD to gain "parity" with legacy versions. Still a problem for cookies+localstorage set with FPI though which is the deal breaker for me, i.e all cookies need to be denied by default.

**Comment by Thorin-Oakenpants** [comment]
Hmmm .. I block all cookies by default (I actually only allow around 8 domains as 1st party, that's it). This also blocks sites from local dom storage. I wonder if this is also blocking indexedDB .. because after almost a month of having indexedDB enabled, I have yet to get ANYTHING put in there from web content - and I visited sites like YT, NYT, whatever .. and I get absolutely no data in my /storage/default+temporary folders .. not a bean .. zip, zilch, zero, nuttin', nothing, nada.

Might do some more testing in a nilla

**Comment by Thorin-Oakenpants** [comment]
Offline Storage: I knew that was dodgy, I think Tom told me - anyway : https://bugzilla.mozilla.org/show_bug.cgi?id=1077986

**Comment by earthlng** [comment]
Your problem with cookies + localstorage only arises because you want to permanently keep those 8 cookies. If you could live without that, the easiest solution would be to allow cookies in FF, optionally block them with uMatrix and clear cookies on shutdown (which also clears localstorage). That's at least what I do. It also has the advantage that I don't need to tamper with FF's site exceptions/permissions which IMO suck anyway because there's no easy way to see all the sites for which a permission was set/changed (afaik). And it would solve the problems that a bunch of people reported here because cookies were/are blocked by our template. Just my 2 cents

**Comment by Thorin-Oakenpants** [comment]
The removal of about:permissions was a blow. We used to be able to see that info

There will be different methods that work for each of us. Me, I'm a lone wolf, zero access to my lair/machine - I like my auto-login for 5 sites (and 3 cookies for site prefs) - these 8 sites I am ok with keeping. There is no 3rd party leakage.

The problem is most of the persistent storage APIs are broken. Clearing history has issues, IDB is a nightmare, and so on. WHEN these are fixed, there should be multiple solutions that work for everyone

**Comment by Thorin-Oakenpants** [comment]
PS: I look up my site permission cookie exceptions by looking in Options>Privacy>Show Cookies

**Comment by curiosity-seeker** [comment]
> I've had indexeddb on for about 10 days I think, yet to see ANYTHING land on my hdd. That said, you can trick sites by allowing indexedDB and denying permissions to the folders, and this then makes them work in some cases, IDK exactly - read https://www.wilderssecurity.com/threads/firefox-and-indexeddb.395113/ first comment

One example is https://www.gmx.net. That site ceates a new entry in ...storage/default. 

Denying write access to the whole default folder does not work unfortunately as several webextensions save data therein (moz-extension+++ .... subnfolders). There are also other subfolders (about+newtab ...) possibly related to containers ?

Anyway, my current strategy is the following:
1. Delete all subfolders in _default_ pointing to specific domains.
2. Go up to the _storage_ folder and remove write permission for _default_:
`chmod -R ug-w default`
3. Add write permission for all existing subfolders in _default_ beginning with "moz" and "about":
```
find ./default/ -name "moz*" -exec chmod -R ug+w {} +
find ./default/ -name "about*" -exec chmod -R ug+w {} +
```
These are Linux commands. I don't know if anything similar is possible for NTFS permissions (but it can certainly be done manually). This works well for me. One disadvantage is that newly installed add-ons are not allowed to save anything in storage. In this case write permission must be re-added for default and later on above steps must be repeated. 

Let's hope that the Firefox devs come up with a better solution. Unfortunately deleting website data in about:preferences#privacy also deletes **all** cookies - including the ones you specifically allowed for, e.g., searx.me or startpage in order to save your search settings.

**Comment by curiosity-seeker** [comment]
Okay - forget what I wrote in the previous post: The issue is fixed in FF58 Nightly which I got today. Obviously related to the fixes mentioned by Pants [here](https://github.com/ghacksuserjs/ghacks-user.js/issues/239#issuecomment-331410098).

**Comment by Thorin-Oakenpants** [comment]
^^ https://bugzilla.mozilla.org/show_bug.cgi?id=1047098 - pushed to 56. TOld ya one of them was a block. This will also make it to ESR I believe

**Comment by atomGit** [comment]
IndexDB - /storage/default

what i'm seeing is that, if you set uM to the global scope and block "other" in the "all" row and then visit something like YouTube, no local storage object gets created (even is allowed in FF permissions UI (default))

to me this is a bit odd because there are no "other" entries in uM for youtube.com - i don't know why uM doesn't handle IndexDB storage better - even if you block cookies in the global scope, the local storage still gets created, as well as cookies - i understand uM doesn't block cookies by default, but rather just prevents them from leaving the browser, but still, if you _purposely_ set it to block then it damn well should

as for automating the removal of local storage objects, SDC is dead and the only other viable alt i'm seeing is [Cookie AutoDelete](https://addons.mozilla.org/en-US/firefox/addon/cookie-autodelete/) (webext), but it doesn't yet handle local storage

[Cookies Exterminator](https://addons.mozilla.org/en-US/firefox/addon/cookies-exterminator/) may work, but it's legacy

**Comment by Thorin-Oakenpants** [comment]
looking at the js

/*** 2701.... network.cookie.cookieBehavior
 * [SETTING] Options>Privacy>History>Custom Settings>Accept cookies from sites
 * [NOTE] This also controls access to 3rd party Web Storage, IndexedDB, Cache API and Service Worker Cache

It certainly controls 1st party AS WELL (for now) for local dom (web storage) and IDB - not sure about two the cache items. This even blocks IDB for extensions. **this* is why I have never seen an IDB entry - because I block all cookies by default - I allow about 12 sites in total now (6 for logins and 6 for breakage or I want DDG to stay in dark mode etc) and none of those sites use IDB.

FF56 clearing Offline Website Data on close and in Delete Recent History (ctrl-Shift-Del) has been fixed. Offline Website Data = IDB entries - this will not clear any web-extension data. I am not 100% sure on the site permissions for "Offline Website Data" yet, and the settings for asking for permission and the default displayed - there are still tickets in for these.

So - block all cookies and use firefox's built in site exceptions (including on web extensions if required). If that's too much work, then Cookei Auto-Delete will be able to clear local dom by host and by time in FF58 (bug landed a few days ago), but I do not know about when an extension will be able to clear IDB, but it is planned. Otherwise you can flick open sites in a PB window, and clear IDB on close. FPI helps (but so does the cookie setting AFAIK) from 3rd parties, but we block 3rd parties. FPI does not protect you from the first party (eg on a revisit)

Whichever way you slice and dice it mate - default deny cookies has always been the best strategy. If I can add working site exceptions for IDB on extensions, I'm killing IDB again :)

**Comment by atomGit** [comment]
> default deny cookies has always been the best strategy

how are you whitelisting extensions so they can use local storage? one by one or a uM rule?

**Comment by Thorin-Oakenpants** [comment]
> how are you whitelisting extensions so they can use [edit by thorin: cookies + local dom + IDB]

https://github.com/ghacksuserjs/ghacks-user.js/issues/250

**Comment by atomGit** [comment]
that's a pain though - one more thing to have to remember when you install an ext. - whatever works for you though

for me, i think i'll stick with allowing 1st party cookies in FF, "blocking" them with uM, and blocking "other" globally in uM - i have nothing in /storage/default other than extension stuff with that config

**Comment by Thorin-Oakenpants** [comment]
> that's a pain though - one more thing to have to remember when you install an ext

My set of extensions is uber stable - its just this migration to web extensions. No big deal IMO - for those that are willing to do the tiny bit of legwork, I feel this is far superior to allowing cookies  + localdom + IDB for all 1st party and then auto-clearing it - do you remember my 1.5million uBo blocks .. imagine that in disk writes over time - its just extra cpu work and performance drag :)

Each to their own. Both strategies work :) especially when extensions can properly control ALL the persistent storage by host and by time

**Comment by Thorin-Oakenpants** [comment]
closing .. so much has happened since this issue was opened over 2 months ago. Both our understanding and fixes and tickets yet to land that have some real traction, and extensions

see #235 for further comments / discussion on a persistent [web content] storage section in the user.js

**Comment by Ratator** [comment]
fsd
